src/mlir/generator.cc

```diff
@@ -716,36 +697,162 @@ namespace mlir::verona
     return ReturnValue();
   }
 
-  llvm::Expected<ReturnValue> Generator::genOperation(
+  // ===================================================== MLIR Generators
+  llvm::Expected<mlir::FuncOp> Generator::generateProto(
+    mlir::Location loc,
+    llvm::StringRef name,
+    llvm::ArrayRef<mlir::Type> types,
+    llvm::ArrayRef<mlir::Type> retTy)
+  {
+    assert(!functionTable.inScope(name) && "Duplicated function declaration");
+
+    // Create function
+    auto funcTy = builder.getFunctionType(types, retTy);
+    auto func = mlir::FuncOp::create(loc, name, funcTy);
+    functionTable.insert(name, func);
+    return func;
+  }
+
+  llvm::Expected<mlir::FuncOp> Generator::generateEmptyFunction(
+    mlir::Location loc,
+    llvm::StringRef name,
+    llvm::ArrayRef<llvm::StringRef> args,
+    llvm::ArrayRef<mlir::Type> types,
+    llvm::ArrayRef<mlir::Type> retTy)
+  {
+    assert(args.size() == types.size() && "Argument/type mismatch");
+
+    // If it's not declared yet, do so. This simplifies direct declaration of
+    // compiler functions. User functions should be checked at the parse level.
+    if (!functionTable.inScope(name))
+    {
+      auto proto = generateProto(loc, name, types, retTy);
+      if (auto err = proto.takeError())
+        return std::move(err);
+      functionTable.insert(name, *proto);
+    }
+    auto func = functionTable.lookup(name);
+
+    // Create entry block, set builder entry point
+    auto& entryBlock = *func.addEntryBlock();
+    auto argVals = entryBlock.getArguments();
+    assert(args.size() == argVals.size() && "Argument/value mismatch");
+    builder.setInsertionPointToStart(&entryBlock);
+
+    // Declare all arguments
+    for (auto var_val : llvm::zip(args, argVals))
+    {
+      // Get the argument name/value
+      auto name = std::get<0>(var_val);
+      auto value = std::get<1>(var_val);
+      // Allocate space in the stack & store the argument value
+      auto alloca = generateAlloca(loc);
+      auto store = generateStore(loc, value, alloca);
+      // Associate the name with the alloca SSA value
+      declareVariable(name, alloca);
+    }
+
+    return func;
+  }
+
+  llvm::Error Generator::generateCondBranch(
+    mlir::Location loc,
+    mlir::Value cond,
+    mlir::Block* ifBB,
+    mlir::ValueRange ifArgs,
+    mlir::Block* elseBB,
+    mlir::ValueRange elseArgs)
+  {
+    // Cast to i1 if necessary
+    auto cast = generateAutoCast(loc, cond, builder.getI1Type());
+    builder.create<mlir::CondBranchOp>(
+      loc, cast, ifBB, ifArgs, elseBB, elseArgs);
+    return llvm::Error::success();
+  }
+
+  llvm::Error
+  Generator::generateLoopBranch(mlir::Location loc, llvm::StringRef blockName)
+  {
+    if (!loopTable.inScope(blockName))
+      return parsingError("Loop branch without a loop", loc);
+    auto block = loopTable.lookup(blockName);
+    mlir::ValueRange empty{};
+    // We assume the branch is the last operation in its basic block
```

Comment:
Can we check this assumption in debug?

Reply:
We check this by running the `mlir::verify()` function at the end of the module generation, on both debug and release builds.